From 9ee77f8b79ca1e449690b637467c3eda141c5ebb Mon Sep 17 00:00:00 2001
From: Razvan Cojocaru <rcojocaru@bitdefender.com>
Date: Mon, 7 Mar 2016 17:09:13 +0200
Subject: [PATCH 1/3] xen: Make sure interrupts don't interfere with the
 re-execution mechanism

vmx_intr_assist() returns early if v->arch.hvm_vcpu.single_step, but
doesn't if v->arch.hvm_vmx.exec_control has already set
CPU_BASED_MONITOR_TRAP_FLAG and v->arch.hvm_vcpu.single_step == 0,
which messes with the re-execution mechanism. This patch sets
(and restores) v->arch.hvm_vcpu.single_step instead of just MTF.

Signed-off-by: Razvan Cojocaru <rcojocaru@bitdefender.com>
---
 xen/arch/x86/hvm/vmx/vmx.c   | 1 +
 xen/arch/x86/mm/p2m.c        | 6 ++----
 xen/include/asm-x86/domain.h | 1 +
 3 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/xen/arch/x86/hvm/vmx/vmx.c b/xen/arch/x86/hvm/vmx/vmx.c
index de26f43..18cfe77 100644
--- a/xen/arch/x86/hvm/vmx/vmx.c
+++ b/xen/arch/x86/hvm/vmx/vmx.c
@@ -2936,6 +2936,7 @@ static int vmx_stop_reexecute_instruction(struct vcpu *v)
         }
 
         v->arch.rexec_context[i].gpa = 0;
+        v->arch.hvm_vcpu.single_step = v->arch.rexec_context[i].old_single_step;
     }
 
     spin_lock(&v->domain->arch.rexec_lock);
diff --git a/xen/arch/x86/mm/p2m.c b/xen/arch/x86/mm/p2m.c
index aa683d0..7c683fb 100644
--- a/xen/arch/x86/mm/p2m.c
+++ b/xen/arch/x86/mm/p2m.c
@@ -1775,6 +1775,7 @@ release_and_exit:
 
     v->arch.rexec_context[level].gpa = gpa;
     v->arch.rexec_context[level].old_access = old_access;
+    v->arch.rexec_context[level].old_single_step = v->arch.hvm_vcpu.single_step;
 
     /* Step 5: Make the GPA with the required access, so we can re-execute
      * the instruction. */
@@ -1822,10 +1823,7 @@ release_and_exit:
      * VM-exit to be generated after the instruction has been
      * successfully re-executed. */
     if ( 0 == level )
-    {
-        v->arch.hvm_vmx.exec_control |= CPU_BASED_MONITOR_TRAP_FLAG;
-        vmx_update_cpu_exec_control(v);
-    }
+        v->arch.hvm_vcpu.single_step = 1;
 
     /* Step 8: We should be done! */
 
diff --git a/xen/include/asm-x86/domain.h b/xen/include/asm-x86/domain.h
index f183595..0c75724 100644
--- a/xen/include/asm-x86/domain.h
+++ b/xen/include/asm-x86/domain.h
@@ -570,6 +570,7 @@ struct arch_vcpu
         unsigned long gpa;
         xenmem_access_t old_access;
         xenmem_access_t cur_access;
+        bool_t old_single_step;
     } rexec_context[REEXECUTION_MAX_DEPTH];
 
     int rexec_level;
-- 
2.7.3

